sql/core/src/main/scala/org/apache/spark/sql/execution/BufferedRowIterator.java

```diff
@@ -64,7 +66,16 @@ protected void append(InternalRow row) {
    * If it returns true, the caller should exit the loop (return from processNext()).
    */
   protected boolean shouldStop() {
-    return !currentRows.isEmpty();
+    return !currentRows.isEmpty() || stopEarly;
```

Comment:
I think we could only override `shouldStop()` for Limit, so we won't add a branch when we don't have Limit.

```
ctx.addFunction("shouldStop", """
  protected boolean shouldStop() {
    return !currentRows.isEmpty() || stopEarly;
 }""")
```

Please be care full that we can't have two `shouldStop()` in generated class.